DSEGOG-101 specify max shots to display

sam-glendenning opened the pull request:
## Description
This adds radio buttons to allow the user to specify the maximum number of shots
- to display in the plot
- to paginate through in the record table

The current setting for max shots is stored as part of the search parameters, meaning things only update when the user clicks Search.

## Testing instructions
There's only so much testing of this that can be done considering the options for Max Shots are 50, 1000 and Unlimited while the test data we have is limited to just over 100 shots. However, I've mocked things where I can in tests. At the moment, when you click on 1000 or unlimited, the correct logic should kick in so that the table and plots display all shots.

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking
Closes DSEGOG-101

Commit:
DSEGOG-101 checkpoint for initial max shots popup

Adding popup to live alongside other search components

Commit:
Merge branch 'develop' into DSEGOG-101-max-shots

Commit:
DSEGOG-101 adding max shots functionality

This limits the total rows available to display in the record table (without altering the backend query). The max shots number is currently in the search params store but because it's not actually part of the request I may change this later on

Commit:
DSEGOG-101 change search button style on param change

Whenever the search parameters are changed, the search button style is changed to indicate this has happened and that a search on new parameters can be performed. This carries over from eCat

Commit:
DSEGOG-101 add max shots unit tests

Commit:
DSEGOG-101 adding unit tests for max shots selector

Still need to work on the placement and styling of the component

Commit:
DSEGOG-101 usePlotRecords leverages maxShots

The plots now only display the maximum number of shots selected in the data view. I did this by imposing skip and limit params to make it act like it's fetching one giant page of data

Commit:
DSEGOG-101 updating maxShots and records tests

Updating usePlotRecords test to reflect addition of maxShots param to backend query

Commit:
Merge branch 'develop' into DSEGOG-101-max-shots

Commit:
DSEGOG-101 adding e2e tests

Stub the counts to match what maxShots equals

Commit:
DSEGOG-101 improving position of maxShots component

Now underneath the rest of the search bar in a row of options

Commit:
DSEGOG-101 adding and updating table tests

Adding missing table tests

Adding tests for checking the data count for tables

Commit:
DSEGOG-101 improving maxShots layou

Allows for different screen sizes to render the component nicely

Commit:
DSEGOG-101 increasing SLOPPY_CLICK_THRESHOLD for column moving e2e tests

This accounts for the table being shifted down the page slightly following the addition of the max shots component

louise-davies commented on src/app.types.tsx:
```diff
@@ -124,6 +124,7 @@ export interface ShotnumRange {
 export interface SearchParams {
   dateRange: DateRange;
   shotnumRange: ShotnumRange;
+  maxShots: number | 'Unlimited';
```
I wonder if some of the logic here would be simplified if "Unlimited" `maxShots` is represented by `Infinity`. It would mean no special handling for the table - as `totalDataCount < maxShots ? totalDataCount : maxShots` works if `maxShots` is `Infinity`. Also means we don't have to deal with both numbers and strings in parsing the values in `maxShots.component.tsx` - the only change is instead of using `parseInt(value)` we'd have to use `Number(value)`. Thoughts? Feel free to ignore if you try it and the logic is more complex

sam-glendenning replied:
Use of `Infinity` had not crossed my mind but I like it and it does simplify things. I'll make the change

louise-davies reviewed (commented):
Tested it and it's working fine with both the table and plots. Just my comment on using `Infinity` instead of a string and I'm happy for it to go in :)

Commit:
DSEGOG-101 updating Unlimited value to Infinity

Allows for easier parsing of the maxShots value

Commit:
Merge branch 'develop' into DSEGOG-101-max-shots

sam-glendenning commented:
There are definitely conflicts between this and https://github.com/ral-facilities/operationsgateway/pull/138 so approve one before the other

Commit:
DSEGOG-101 adding aria label to max shots radio buttons

This allows for better e2e testing (otherwise, we would have to find an input with a value of Infinity, and Infinity is not represented as a value on the screen, as opposed to Unlimited)

This also improves accessibility for users with screen leaders to click the correct option

Commit:
Merge branch 'develop' into DSEGOG-101-max-shots

Commit:
DSEGOG-101 fixing shotnum range bug

An invalid range check was erroneously parsing shotnum values of 1 and 0 as their boolean equivalents, instead of correctly checking if those values were defined or not. This has been fixed and the unit test for it has been updated to check for this

Commit:
DSEGOG-101 fixing unit tests with added aria label
